packages/react-reconciler/src/ReactFiberNewContext.js

```diff
@@ -90,17 +97,178 @@ function popProvider(providerFiber: Fiber): void {
+    let nextFiber;
+
+    // Visit this fiber.
+    let reader = fiber.firstContextReader;
+    if (reader !== null) {
+      do {
+        // Check if the context matches.
+        if (
+          reader.context === context &&
+          (reader.observedBits & changedBits) !== 0
+        ) {
+          // Match! Update the expiration time of all the ancestors, including
+          // the alternates.
+          let node = fiber;
+          while (node !== null) {
+            const alternate = node.alternate;
+            if (
+              node.expirationTime === NoWork ||
+              node.expirationTime > renderExpirationTime
+            ) {
+              node.expirationTime = renderExpirationTime;
+              if (
+                alternate !== null &&
+                (alternate.expirationTime === NoWork ||
+                  alternate.expirationTime > renderExpirationTime)
+              ) {
+                alternate.expirationTime = renderExpirationTime;
+              }
+            } else if (
+              alternate !== null &&
+              (alternate.expirationTime === NoWork ||
+                alternate.expirationTime > renderExpirationTime)
+            ) {
+              alternate.expirationTime = renderExpirationTime;
+            } else {
+              // Neither alternate was updated, which means the rest of the
+              // ancestor path already has sufficient priority.
+              break;
+            }
+            node = node.return;
+          }
+          // Don't scan deeper than a matching consumer. When we render the
+          // consumer, we'll continue scanning from that point. This way the
+          // scanning work is time-sliced.
+          nextFiber = null;
+        } else {
+          nextFiber = fiber.child;
+        }
+        reader = reader.next;
+      } while (reader !== null);
+    } else if (fiber.tag === ContextProvider) {
+      // Don't scan deeper if this is a matching provider
+      nextFiber = fiber.type === workInProgress.type ? null : fiber.child;
+    } else {
+      // Traverse down.
+      nextFiber = fiber.child;
+    }
+
+    if (nextFiber !== null) {
+      // Set the return pointer of the child to the work-in-progress fiber.
+      nextFiber.return = fiber;
+    } else {
+      // No child. Traverse to next sibling.
+      nextFiber = fiber;
+      while (nextFiber !== null) {
+        if (nextFiber === workInProgress) {
+          // We're back to the root of this subtree. Exit.
+          nextFiber = null;
+          break;
+        }
+        let sibling = nextFiber.sibling;
+        if (sibling !== null) {
+          // Set the return pointer of the sibling to the work-in-progress fiber.
+          sibling.return = nextFiber.return;
+          nextFiber = sibling;
+          break;
+        }
+        // No more siblings. Traverse up.
+        nextFiber = nextFiber.return;
+      }
+    }
+    fiber = nextFiber;
+  }
+}
+
+export function checkForPendingContext(
```

Comment:
This function is way too slow to iterate through these every time we rerender even if we're about to just rerender anyway. We'll need a different approach. 

This should already have been known by the context propagation so we just need some way of storing that.

Reply:
What about the part where we continue the propagation where we left off?

Reply:
We could do that part only if any of the contexts matched. The point is to optimize for the case when zero context changed.